### ISSUE TITLE
Add support for setting advertising data

### ISSUE THREAD

**Issue by jockej** [intel-iot-devkit/tinyb#136]
I need to be able to set the advertising data for a project I'm working on.
This seems to be possible with Bluez, as per here: https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc/advertising-api.txt
but I could not find a method to do it in tinyb (I'm working in Java).

I could give it a go, but I would need some pointers (:-)) on how to do it.

Or maybe somebody else feels like having a try :)

**Comment by mburger81** [comment]
Hi @jockej have a look into this issue, #49 

Was you able to implement BLE advertising? We try to do the same thing.

thx

**Comment by jockej** [comment]
As far as I understood #49 i solved.

This is not really the same thing, I believe. I want to set the advertising data when advertising. Not read advertisement data from other devices.

**Comment by mburger81** [comment]
You wrote "I solved" so you intend you solved your requirement? Can you help?

**Comment by lachev** [comment]
Hi there, I'm trying also - as for tinyb - from dbus - c code c++, JNI and up to java. Made my functions and so - but on register _advertisement - it is ok and one instance is used as created LEAdvertiseingManager and LEAdvertisement - but some more stuff must do -  leadvertising_manager1_call_register_advertisement_sync returns ok and get_active_instances become 1 - but I cannot sniff them from my other BLE device. With pyton script in examples is ok (example-advertisement). I'm using BlueZ
 5,48 in experimental

**Comment by lachev** [comment]
Note that: for <interface name="org.bluez.LEAdvertisement1"> ; <property name="ManufacturerData" type="a{qv}" access="read"/>  ---- a(qv) not a(sv) for my point of view

**Comment by lachev** [comment]
for:
  <interface name="org.bluez.LEAdvertisingManager1">
    <method name="RegisterAdvertisement">
      <arg name="advertisement" type="o" direction="in"/>
      <arg name="options" type="a{sv}" direction="in"/>
    </method>
    <method name="UnregisterAdvertisement">
      <arg name="service" type="o" direction="in"/>
    </method>
    <property name="ActiveInstances" type="y" access="read"/>
    <property name="SupportedInstances" type="y" access="read"/>
    <property name="SupportedIncludes" type="as" access="read"/>
  </interface>
the properties I've got working but on RegisterAdvertisment it says ok but nothing is sending - as far i can see here is some HCI stuf for setting some times and other stuff

**Comment by mburger81** [comment]
@lachev WHAT? :) sorry but your message is cryptic for me :)
Where is the LEAdvertisingManager?

**Comment by jockej** [comment]
@mburger81 Sorry, I meant "is solved". The last word on that thread seems to be that they got it working.

**Comment by lachev** [comment]
when you see /lib/systemd/system/bluetooth.service - inside must be: ExecStart=/usr/local/libexec/bluetooth/bluetoothd --experimental and then in d-feet uoy will see this interface

**Comment by lachev** [comment]
<interface name="org.bluez.LEAdvertisingManager1">
    <method name="RegisterAdvertisement">
      <arg name="advertisement" type="o" direction="in"/>
      <arg name="options" type="a{sv}" direction="in"/>
    </method>
    <method name="UnregisterAdvertisement">
      <arg name="service" type="o" direction="in"/>
    </method>
    <property name="ActiveInstances" type="y" access="read"/>
    <property name="SupportedInstances" type="y" access="read"/>
    <property name="SupportedIncludes" type="as" access="read"/>
  </interface>

**Comment by mburger81** [comment]
Ok but that's not implemented in tinyb, right?

**Comment by lachev** [comment]
[LEAdvertisingManager1.txt](https://github.com/intel-iot-devkit/tinyb/files/1659986/LEAdvertisingManager1.txt)
this is the interface that was unable to be seen

**Comment by lachev** [comment]
yes there is no code in tinyb except for auto generated one for LEAdvertisement and nothing for LEAddvertisingManager. Trying to find some solution from 2 weeks

**Comment by jockej** [comment]
@lachev Have you tried to implement this? Do you have a fork on github?

**Comment by lachev** [comment]
I have a fork – and I feel that I’m going closer – ASAP I have clear code I’ll do it.

Sent from Mail for Windows 10

**Comment by mburger81** [comment]
@lachev beatufil, if you like we can try your fork.

**Comment by lachev** [comment]
This days it is full of printouts and commented code


Sent from Mail for Windows 10

**Comment by lachev** [comment]
You can check this https://github.com/carsonmcdonald/bluez-ibeacon (inside bluez-beacon - C written). It is simple and can be added somehow. It seems for me that HCI layer is very important for advertisements. There are some timings and so on. I cannot see it right now over DBUS but can be added like some properties or else.


Sent from Mail for Windows 10
